src/graphql/graphql-api.php

```diff
@@ -0,0 +1,245 @@
+		if ( isset( $filter_options['graphQL'] ) && $filter_options['graphQL'] ) {
+
+			// Add the ID to the block, if it is not already there.
+			if ( ! isset( $sourced_block['id'] ) && isset( $filter_options['id'] ) ) {
+				$sourced_block['id'] = $filter_options['id'];
+			}
+			
+			// Flatten the inner blocks, if any.
+			if ( isset( $sourced_block['innerBlocks'] ) ) {
+				$sourced_block['innerBlocks'] = self::flatten_inner_blocks( $sourced_block['innerBlocks'], $filter_options );
+			}
+
+			// Convert the attributes to be in the name-value format that the schema expects.
+			if ( isset( $sourced_block['attributes'] ) && ! empty( $sourced_block['attributes'] ) && ! isset( $sourced_block['attributes'][0]['name'] ) ) {
+				$sourced_block['attributes'] = array_map(
+					function ( $name, $value ) {
+						return [
+							'name'  => $name,
+							'value' => $value,
+						];
+					},
+					array_keys( $sourced_block['attributes'] ),
+					array_values( $sourced_block['attributes'] )
+				);
+			}
+		}
+
+		return $sourced_block;
+	}
+
+	/**
+	 * Flatten the inner blocks, no matter how many levels of nesting is there.
+	 *
+	 * @param array $inner_blocks the inner blocks in the block.
+	 * @param array $filter_options Options to filter using, if any.
+	 * @param array $flattened_blocks the flattened blocks that's built up as we go through the inner blocks.
+	 * 
+	 * @return array
+	 */
+	public static function flatten_inner_blocks( $inner_blocks, $filter_options, $flattened_blocks = [] ) {
+		foreach ( $inner_blocks as $inner_block ) {
+			// This block doesnt have any inner blocks, so just add it to the flattened blocks. Ensure the parentId is set.
+			if ( ! isset( $inner_block['innerBlocks'] ) ) {
+				$inner_block['parentId'] = $inner_block['parentId'] ?? $filter_options['parentId'];
+				array_push( $flattened_blocks, $inner_block );
+				// This block is a root block, so go through the inner blocks recursively.
+			} elseif ( ! isset( $inner_block['parentId'] ) ) {
+				$inner_blocks_copy = $inner_block['innerBlocks'];
+				unset( $inner_block['innerBlocks'] );
+				$inner_block['parentId'] = $inner_block['parentId'] ?? $filter_options['parentId'];
+				array_push( $flattened_blocks, $inner_block );
+				$flattened_blocks = self::flatten_inner_blocks( $inner_blocks_copy, $filter_options, $flattened_blocks );
+			}
+		}
+
+		return $flattened_blocks;
+	}
+
+	/**
+	 * Register types and fields graphql integration.
+	 *
+	 * @return void
+	 */
+	public static function register_types() {
+		// Register the type corresponding to the attributes of each individual block.
+		register_graphql_object_type(
+			'BlockDataAttribute',
+			[
+				'description' => 'Block data attribute',
+				'fields'      => [
+					'name'  => [
+						'type'        => 'String',
```

Comment:
It's good practice to designate fields as non-nullable whenever possible. That way the consumer can have confidence that a field will always have a value and can skip inspection.
```suggestion
						'type'        => [ 'non_null' => 'String' ],
```

Reply:
Good point, I have done this now